### ISSUE TITLE
a negligence of the viterbi algorithm 

### ISSUE THREAD

**Issue by cosmosZhou** [keras-team/keras-contrib#343]
I write my own viterbi algorithm myself, using the same weights trained from the CRF model, I found that the final results from the model.predict and my own viterbi algorithm are sometimes different slightly. Below are my viterbi algorith. 

import numpy as np
class CRF:
    def __init__(self, kernel, chain_kernel, bias, left_boundary, right_boundary):
        self.bias = bias
        self.chain_kernel = chain_kernel
        self.kernel = kernel
        self.left_boundary = left_boundary
        self.right_boundary = right_boundary
        self.activation = linear

    def viterbi_one_hot(self, X):
        return np.eye(len(self.bias), dtype=int)[self.viterbi_decoding(X)]

    def viterbi_decoding(self, X):
        input_energy = self.activation(np.matmul(X, self.kernel) + self.bias)
        input_energy[0] += self.left_boundary
        input_energy[-1] += self.right_boundary
        min_energy = np.zeros_like(input_energy[0])
        argmin_tables = np.zeros(input_energy.shape, int)
        for t in range(len(input_energy)):
            energy = self.chain_kernel + np.expand_dims(input_energy[t] + min_energy, 1)
            argmin_tables[t] = np.argmin(energy, 0)
            min_energy = np.min(energy, 0)

        best_idx = np.argmin(min_energy)
#         best_idx = argmin_tables[-1, 0] # this is what the keras.contrib is doing!!!
        best_paths = np.zeros((len(argmin_tables),), int)
        for i in range(-1, -len(argmin_tables) - 1, -1):
            best_idx = argmin_tables[i][best_idx]
            best_paths[i] = best_idx

        return best_paths

I examine in detail the source codes of CRF model, this is what I found in keras.contrib:

    def viterbi_decoding(self, X, mask=None):
        input_energy = self.activation(K.dot(X, self.kernel) + self.bias)
        if self.use_boundary:
            input_energy = self.add_boundary_energy(input_energy, mask, self.left_boundary, self.right_boundary)

        argmin_tables, argmin_table_last = self.recursion(input_energy, mask, return_logZ=False)
        argmin_tables = K.cast(argmin_tables, 'int32')

        # backward to find best path, `initial_best_idx` can be any, as all elements in the last argmin_table are the same
        argmin_tables = K.reverse(argmin_tables, 1)
         initial_best_idx = [K.expand_dims(argmin_tables[:, 0, 0])]  # matrix instead of vector is required by tf `K.rnn`
        if K.backend() == 'theano':
            initial_best_idx = [K.T.unbroadcast(initial_best_idx[0], 1)]

        def gather_each_row(params, indices):
            n = K.shape(indices)[0]
            if K.backend() == 'theano':
                return params[K.T.arange(n), indices]
            else:
                indices = K.transpose(K.stack([K.tf.range(n), indices]))
                return K.tf.gather_nd(params, indices)

        def find_path(argmin_table, best_idx):
            next_best_idx = gather_each_row(argmin_table, best_idx[0][:, 0])
            next_best_idx = K.expand_dims(next_best_idx)
            if K.backend() == 'theano':
                next_best_idx = K.T.unbroadcast(next_best_idx, 1)
            return next_best_idx, [next_best_idx]

        _, best_paths, _ = K.rnn(find_path, argmin_tables, initial_best_idx, input_length=K.int_shape(X)[1], unroll=self.unroll)
        best_paths = K.reverse(best_paths, 1)
        best_paths = K.squeeze(best_paths, 2)

        return K.tf.one_hot(best_paths, self.units, dtype=K.floatx())

I think there might be something wrong in the sentence above, which I repeat below again:
        # backward to find best path, `initial_best_idx` can be any, as all elements in the last argmin_table are the same
        .......
         initial_best_idx = [K.expand_dims(argmin_tables[:, 0, 0])]  # matrix instead of vector is required by tf `K.rnn`

I think the  initial_best_idx should be selected from the final state, since NOT all elements in the last argmin_table are ALWAYS the same!

**Comment by qiaojiim** [comment]
Any related explanations? I have the same question.
